kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java

```diff
@@ -76,24 +77,27 @@ public FileSystemViewKeyInputFormat(FileSystemDataset<E> dataset,
     Format format = dataset.getDescriptor().getFormat();
 
     boolean isSpecific = SpecificRecord.class.isAssignableFrom(dataset.getType());
+    String readerSchema = conf.get("kite.readerSchema");
+    if (isSpecific && readerSchema == null) {
```

Comment:
I think we currently default the reader schema to the writer schema for specific and reflect (see [`DataModelUtil`](https://github.com/kite-sdk/kite/blob/master/kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/DataModelUtil.java#L167)). That logic is correct for specific and generic, but reflect will need some special handling because there are two ways to interpret the records.

When reading, we should use the dataset's schema, but make sure that it is compatible with the reflected schema using AllowNull because that is what the objects can hold. When writing, we need to use the dataset's schema but validate that it can read the reflected schema without AllowNull or else we will have a problem.